### ISSUE TITLE
How check if the command succeeded?

### ISSUE THREAD

**Issue by jclsn** [tpope/vim-dispatch#339]
Is there any way to see if the build (or any other command) succeeded? It is not clear to me from the readme. I would like to use ``:Make`` to build and start debugging afterwards, but only when the build was successful of course.

**Comment by jclsn** [comment]
Maybe I should be more clear. This is what I am trying

```
function! s:BuildAndDebug()
	:Make

	if dispatch#request().status
		autocmd QuickFixCmdPost make ++once call vimspector#Launch()
	endif
endfunction
```

Can you tell me what I am doing wrong?

**Comment by tpope** [comment]
`:Make` is asynchronous. You're checking the exit status before it has exited.

**Comment by jclsn** [comment]
Right, I could have though of that. I got an output now, but there is nothing like "status". There is "completed" though, but that is always 1, if the build fails or not. This is shown as success or failed though.
```
:!cmake --build build -- (job/36615)
{'pid': 36615, 'background': 1, 'action': 'make', 'job': 'process 36615 run', 'program': 'cmake --build build --', 'command': 'cmake --build build --', 'file': '/tmp/vLZNAVp/67', 'id': 63, 'handler': 'job', 'mods': '', 'directory':
 '/home/jan/Workspace/C++/constexpr', 'format': '%*[^"]"%f"%*\D%l: %m,"%f"%*\D%l: %m,%-G%f:%l: (Each undeclared identifier is reported only once,%-G%f:%l: for each function it appears in.),%-GIn file included from %f:%l:%c:,%-GIn f
ile included from %f:%l:%c\,,%-GIn file included from %f:%l:%c,%-GIn file included from %f:%l,%-G%*[ ]from %f:%l:%c,%-G%*[ ]from %f:%l:,%-G%*[ ]from %f:%l\,,%-G%*[ ]from %f:%l,%f:%l:%c:%m,%f(%l):%m,%f:%l:%m,"%f"\, line %l%*\D%c%*[^
 ] %m,%D%*\a[%*\d]: Entering directory %*[`'']%f'',%X%*\a[%*\d]: Leaving directory %*[`'']%f'',%D%*\a: Entering directory %*[`'']%f'',%X%*\a: Leaving directory %*[`'']%f'',%DMaking %*\a in %f,%f|%l| %m', 'expanded': 'cmake --build
build --', 'title': 'make', 'args': ''}
Success: !cmake --build build -- (job/36615)
```

**Comment by tpope** [comment]
> ` 'job': 'process 36615 run'`

That's the job that is running. If the job is running, then it hasn't exited, and there won't be a `status`. You will need to check *during* the autocommand.

**Comment by tpope** [comment]
Oh, Dispatch fires the `QuickFixCmdPost make` event fires twice, once after the job starts, and another after the job finishes. This was to support use cases where cleanup needs to be done before returning control to the user. So you'll need to get a bit more clever than `++once`.

**Comment by jclsn** [comment]
If I don't use once, Vimspector will be started every time I build. I did succeed doing this using this [CMake plugin ](https://github.com/cdelledonne/vim-cmake), but it doesn't kill backround jobs properly. Would also be nice to have a universal solution for all programming languages. Not so easy apparently.

**Comment by tpope** [comment]
> If I don't use once, Vimspector will be started every time I build.

If you don't use `++once`, _the autocommand will run_ every time you build. You can use code in that autocommand to start Vimspector under the right circumstances. For example, you could set a variable before calling `:Make`, and then check for and unset that variable in the autocommand.

**Comment by tpope** [comment]
Another thing I just noticed is you need to use `:Make`, not `:Make!`, to get the autocommand to run on completion.

**Comment by jclsn** [comment]
I tried both. There is no status value ever

**Comment by tpope** [comment]
```vim
command! Debug call BuildAndDebug()

function! BuildAndDebug()
  let g:debug_pending = 1
  Make
endfunction

function! RunVimspectorIfBuilt()
  if !exists('g:debug_pending') || !has_key(dispatch#request(), 'status')
    return
  endif
  unlet! g:debug_pending
  echomsg 'Exited with status ' . dispatch#request().status
endfunction
```

That'll be $200.

**Comment by jclsn** [comment]
There is being printed for me. I checked the message history multiple times.

**Comment by tpope** [comment]
You copied the full example including the `augroup`? I don't know what else to tell you. It definitely works.

**Comment by jclsn** [comment]
Yeah, the whole thing. For me the function always returns because there is no key called status. Weird. Anyway, thanks for trying

**Comment by tpope** [comment]
This is the correct way to get the status:

```vim
let status = +readfile(dispatch#request().file . '.complete', 1)[0]
```

That will fail with ENOENT if the request is running, so change the initial guard to `has_key(dispatch#request(), 'complete')`.

The `status` key was actually added in PR #313, which I had checked out at the time, hence my confusion. I'll waive the $200 charge.

**Comment by jclsn** [comment]
Ah yeah that finally works. Now I can finally rest in peace. I'll recommend you!

**Comment by camilo863** [comment]
> Bien, podría haber pensado en eso. Obtuve un resultado ahora, pero no hay nada como "estado". Sin embargo, está "completado", pero siempre es 1, si la compilación falla o no. Sin embargo, esto se muestra como un éxito o un fracaso.
> 
> ```
> :!cmake --build build -- (job/36615)
> {'pid': 36615, 'background': 1, 'action': 'make', 'job': 'process 36615 run', 'program': 'cmake --build build --', 'command': 'cmake --build build --', 'file': '/tmp/vLZNAVp/67', 'id': 63, 'handler': 'job', 'mods': '', 'directory':
>  '/home/jan/Workspace/C++/constexpr', 'format': '%*[^"]"%f"%*\D%l: %m,"%f"%*\D%l: %m,%-G%f:%l: (Each undeclared identifier is reported only once,%-G%f:%l: for each function it appears in.),%-GIn file included from %f:%l:%c:,%-GIn f
> ile included from %f:%l:%c\,,%-GIn file included from %f:%l:%c,%-GIn file included from %f:%l,%-G%*[ ]from %f:%l:%c,%-G%*[ ]from %f:%l:,%-G%*[ ]from %f:%l\,,%-G%*[ ]from %f:%l,%f:%l:%c:%m,%f(%l):%m,%f:%l:%m,"%f"\, line %l%*\D%c%*[^
>  ] %m,%D%*\a[%*\d]: Entering directory %*[`'']%f'',%X%*\a[%*\d]: Leaving directory %*[`'']%f'',%D%*\a: Entering directory %*[`'']%f'',%X%*\a: Leaving directory %*[`'']%f'',%DMaking %*\a in %f,%f|%l| %m', 'expanded': 'cmake --build
> build --', 'title': 'make', 'args': ''}
> Success: !cmake --build build -- (job/36615)
> ```

https://github.com/tpope/vim-dispatch/issues/339#issuecomment-1490669364
